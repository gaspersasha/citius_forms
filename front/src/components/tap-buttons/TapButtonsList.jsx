import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { INPUT_STATUS } from '~constants';
import s from './styles/tap-buttons.module.sass';

export const isEmpty = (obj) => {
  if (!obj) return true;

  return Object.keys(obj).length === 0 && obj.constructor === Object;
};

const TapButtonsList = ({
  options,
  value: apiVal,
  onChange,
  status,
  defaultVal,
  sectionName,
  ...rest
}) => {
  const isValueExist = (value) => value || value === false || value === 0;
  const labelClass = classNames(s.label, {
    [s.error]: status === INPUT_STATUS.INVALID,
  });

  return (
    <>
      {options.map(({ option, value: elVal, content }) => {
        const active = isValueExist(apiVal)
          ? apiVal === elVal
          : defaultVal === elVal;
        const buttonClass = classNames(s.item, { [s.active]: active });

        const handleChange = (e) =>
          onChange({
            ...e,
            target: { ...e.target, name: e.target.name, value: elVal },
          });

        return (
          <div className={buttonClass} key={elVal}>
            <label htmlFor={option} className={labelClass}>
              <span className={s.labelText}>{option}</span>
              <input
                className={s.input}
                type="radio"
                name={sectionName}
                value={elVal}
                onChange={handleChange}
                checked={active}
                data-id={option || null}
                {...rest}
              />
              <span className={s.checkmark} />
            </label>
            {!isEmpty(content) && <content.Component {...content.props} />}
          </div>
        );
      })}
    </>
  );
};

TapButtonsList.propTypes = {
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.any.isRequired,
      option: PropTypes.string.isRequired,
    })
  ).isRequired,
  value: PropTypes.any,
  defaultVal: PropTypes.any,
  status: PropTypes.string,
  sectionName: PropTypes.string,
};

TapButtonsList.defaultProps = {
  value: '',
  defaultVal: '',
  status: INPUT_STATUS.DEFAULT,
  sectionName: '',
};

export default TapButtonsList;
