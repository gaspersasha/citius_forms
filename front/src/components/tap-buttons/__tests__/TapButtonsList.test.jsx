import React from 'react';
import TapButtonsList from '../TapButtonsList';

const Component = () => <div />;

const defaultProps = {
  options: [
    {
      value: 'TEST0',
      option: 'Test0',
      content: { Component, props: {} },
    },
    {
      value: 'TEST1',
      option: 'Test1',
    },
  ],
  onChange: jest.fn(),
};

describe('<TapButtonsList />', () => {
  let wrapper;

  const getWrapper = (props) =>
    shallow(<TapButtonsList {...defaultProps} {...props} />);

  it('renders without errors', () => {
    wrapper = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders content if it is provided', () => {
    wrapper = getWrapper();
    expect(wrapper.find('Component').exists()).toBe(true);
  });

  it('assigns active option', () => {
    wrapper = getWrapper({ value: defaultProps.options[0].value });
    expect(wrapper.find('input').at(0).prop('checked')).toBe(true);
  });

  it('calls change event calback', () => {
    const name = 'test';

    wrapper = getWrapper();

    wrapper.find('input').at(0).simulate('change', { target: { name } });
    expect(defaultProps.onChange).toBeCalledWith({
      target: { name, value: defaultProps.options[0].value },
    });
  });
});
