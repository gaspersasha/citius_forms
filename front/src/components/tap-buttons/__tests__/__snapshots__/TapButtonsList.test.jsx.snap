// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TapButtonsList /> renders without errors 1`] = `
<Fragment>
  <div
    className="item"
    key="TEST0"
  >
    <label
      className="label"
      htmlFor="Test0"
    >
      <span
        className="labelText"
      >
        Test0
      </span>
      <input
        checked={false}
        className="input"
        data-id="Test0"
        name=""
        onChange={[Function]}
        type="radio"
        value="TEST0"
      />
      <span
        className="checkmark"
      />
    </label>
    <Component />
  </div>
  <div
    className="item"
    key="TEST1"
  >
    <label
      className="label"
      htmlFor="Test1"
    >
      <span
        className="labelText"
      >
        Test1
      </span>
      <input
        checked={false}
        className="input"
        data-id="Test1"
        name=""
        onChange={[Function]}
        type="radio"
        value="TEST1"
      />
      <span
        className="checkmark"
      />
    </label>
  </div>
</Fragment>
`;
